tensorflow/notebooks/sysml_pipeline/6_predictions.ipynb

```diff
@@ -0,0 +1,221 @@
+    "    def throttle(self):\n",
+    "        with self._condition:\n",
+    "            while self._active == self._concurrency:\n",
+    "                self._condition.wait()\n",
+    "            self._active += 1\n",
+    "\n",
+    "\n",
+    "def _create_rpc_callback(label, result_counter):\n",
+    "    \"\"\"Creates RPC callback function.\n",
+    "\n",
+    "    Args:\n",
+    "      label: The correct label for the predicted example.\n",
+    "      result_counter: Counter for the prediction result.\n",
+    "    Returns:\n",
+    "      The callback function.\n",
+    "    \"\"\"\n",
+    "    def _callback(result_future):\n",
+    "        \"\"\"Callback function.\n",
+    "\n",
+    "        Calculates the statistics for the prediction result.\n",
+    "\n",
+    "        Args:\n",
+    "          result_future: Result future of the RPC.\n",
+    "        \"\"\"\n",
+    "        exception = result_future.exception()\n",
+    "        if exception:\n",
+    "            result_counter.inc_error()\n",
+    "            print(exception)\n",
+    "        else:\n",
+    "            sys.stdout.write('.')\n",
+    "            sys.stdout.flush()\n",
+    "            response = numpy.array(\n",
+    "                result_future.result().outputs['scores'].float_val)\n",
+    "            prediction = numpy.argmax(response)\n",
+    "            if label != prediction:\n",
+    "                result_counter.inc_error()\n",
+    "        result_counter.inc_done()\n",
+    "        result_counter.dec_active()\n",
+    "    return _callback\n",
+    "\n",
+    "\n",
+    "def do_inference(hostport, work_dir, concurrency, num_tests):\n",
+    "    \"\"\"Tests PredictionService with concurrent requests.\n",
+    "\n",
+    "    Args:\n",
+    "      hostport: Host:port address of the PredictionService.\n",
+    "      work_dir: The full path of working directory for test data set.\n",
+    "      concurrency: Maximum number of concurrent requests.\n",
+    "      num_tests: Number of test images to use.\n",
+    "\n",
+    "    Returns:\n",
+    "      The classification error rate.\n",
+    "\n",
+    "    Raises:\n",
+    "      IOError: An error occurred processing test data set.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    test_data_set = mnist_input_data.read_data_sets(work_dir).test\n",
+    "    result_counter = _ResultCounter(num_tests, concurrency)\n",
+    "\n",
+    "    futures = [pool.submit(send_single_request,test_data_set) for i in range(num_tests)]\n",
+    "    results = [r.result() for r in as_completed(futures)]\n",
+    "\n",
+    "\n",
+    "def send_single_request(test_data_set):\n",
+    "    image, label = test_data_set.next_batch(1)\n",
+    "    request_data={}\n",
+    "#         request_data['signature_name'] = 'dense_input'\n",
+    "    request_data['instances'] = [[1,2,3,4]]\n",
+    "    headers={}\n",
+    "    headers[constants.HTTP_CONFIG.HTTP_AUTHORIZATION] = \"Bearer \" + util.get_jwt()\n",
```

Comment:
Shouldn't this be refactored to avoid directly getting a jwt token?

Reply:
Yes in hops-util-py we have `util.send_request` method.
@o-alex you did not use the `send_request` due to the random jwt errors you encountered when running the notebooks as jobs?